src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
     }
     n.compile_objects()
 }
+
+fn generate_ffi_mod() {
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").expect("outdir"));
+    let src_dir = env::current_dir().expect("failed to get current directory");
+
+    eprintln!("OUT_DIR={}", out_dir.display());
+    eprintln!("SRC_DIR={}", src_dir.display());
+    let ffi_input = src_dir.join("src").join("ffi.rs");
+    let ffi_src = std::fs::read_to_string(ffi_input.clone()).expect("failed to read src/ffi.rs");
+    println!("{}", format!("cargo:rerun-if-changed={}", ffi_input.display()));
+
+    // Parse src/ffi.rs generating Rust and C wrappers
+    let decls = parse_decls(&ffi_src);
+    let rust_wrapper = generate_rust_wrappers(&decls);
+    let cxx_wrapper = generate_cxx_wrappers(&decls);
+
+    let rust_output = &format!("{}", rust_wrapper.join(""));
+    let rust_ffi_out = out_dir.join("ffi.rs");
+    std::fs::write(rust_ffi_out, rust_output).unwrap();
+
+    let cxx_output = &format!("{}{}", cxx_header(), cxx_wrapper.join(""));
+    let cxx_out = out_dir.join("mozjpeg_cxx_wrapper.cpp");
+    std::fs::write(cxx_out.clone(), cxx_output).unwrap();
+
+    let mozjpeg_include = src_dir.join("vendor");
+    let config_dir = out_dir.join("include");
+    cc::Build::new()
+        .file(cxx_out)
+        .flag("-Wno-return-type-c-linkage")
+        .include(mozjpeg_include)
+        .include(config_dir)
+        .cpp(true)
+        .compile("mozjpeg_cxx_wrapper");
+}
+
+#[derive(Debug)]
+struct Decl {
+    name: String,
+    is_pub: bool,
+    doc: Vec<String>,
+    attrs: Vec<String>,
+    args: Vec<(String, String)>,
+    ret_ty: Option<String>,
+    lifetime: Option<String>
+}
```

Comment:
This is a poor man's version of `syn::ForeignItemFn`, if you just use syn, you get all of this for free.